lib/rexml/node.rb

```diff
@@ -52,10 +52,14 @@ def parent?
 
     # Visit all subnodes of +self+ recursively
     def each_recursive(&block) # :yields: node
-      self.elements.each {|node|
-        block.call(node)
-        node.each_recursive(&block)
-      }
+      stack = []
+      each { |child| stack.unshift child if child.node_type == :element }
+      until stack.empty?
+        child = stack.pop
+        yield child
+        n = stack.size
+        child.each { |grandchild| stack.insert n, grandchild if grandchild.node_type == :element }
+      end
```

Comment:
Can we improve `REXML::Elements#each(nil)` case instead?

Reply:
This seems a possible improvement. But, IMHO, it can be split into another pull request.

Reply:
Do we still need to change `REXML::Node#each_recursive` when we improve `REXML::Elements#each(nil)`?

Reply:
This implementation is best so like that we don't need to change it when we improve something.

I don't really like implementations that are optimized when using specific arguments. From that point of view, I'm not very keen on optimizing `each(nil)`.

Reply:
If `elements.each(nil)` is improved, we can use `elements.each { |x| yield x }` instead of `each { |x| yield x if x.node_type == :element }`.

However, I don't think it is a problem that there are some duplications of `each { |x| yield x if x.node_type == :element }` sometimes. Because `elements.each(&)` has some overheads and the performance is more important here.

Repeatedly, I talked about the **performance**. The maintainability and other matters are out of topic here because the speed of `each_recursive` is one of the crucial parts of the XML library.

Reply:
Thank you for your understanding!